File: JavaScript/Graph/L_modifyGraphEdgeWeight.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @param {number} target
 * @return {number[][]}
 */
var modifiedGraphEdges = function (n, edges, source, destination, target) {
    class PriorityQueue {
        constructor() {
            this.elements = [];
        }

        enqueue(element, priority) {
            this.elements.push({ element, priority });
            this.elements.sort((a, b) => a.priority - b.priority);
        }

        dequeue() {
            return this.elements.shift();
        }

        isEmpty() {
            return this.elements.length === 0;
        }
    }

    const adjacencyList = Array.from({ length: n }, () => []);
    for (let i = 0; i < edges.length; i++) {
        const [nodeA, nodeB] = edges[i];
        adjacencyList[nodeA].push([nodeB, i]);
        adjacencyList[nodeB].push([nodeA, i]);
    }

    const distances = Array.from({ length: n }, () => [Infinity, Infinity]);
    distances[source] = [0, 0];

    runDijkstra(adjacencyList, edges, distances, source, 0, 0);
    const difference = target - distances[destination][0];
    if (difference < 0) return [];

    runDijkstra(adjacencyList, edges, distances, source, difference, 1);
    if (distances[destination][1] < target) return [];

    for (const edge of edges) {
        if (edge[2] === -1) edge[2] = 1;
    }

    return edges;

    function runDijkstra(adjacencyList, edges, distances, source, difference, run) {
        const pq = new PriorityQueue();
        pq.enqueue(source, 0);
        distances[source][run] = 0;

        while (!pq.isEmpty()) {
            const { element: currentNode, priority: currentDistance } = pq.dequeue();

            if (currentDistance > distances[currentNode][run]) continue;

            for (const [nextNode, edgeIndex] of adjacencyList[currentNode]) {
                let weight = edges[edgeIndex][2];
                if (weight === -1) weight = 1;

                if (run === 1 && edges[edgeIndex][2] === -1) {
                    const newWeight =
                        difference + distances[nextNode][0] - distances[currentNode][1];
                    if (newWeight > weight) {
                        edges[edgeIndex][2] = weight = newWeight;
                    }
                }

                if (distances[nextNode][run] > distances[currentNode][run] + weight) {
                    distances[nextNode][run] = distances[currentNode][run] + weight;
                    pq.enqueue(nextNode, distances[nextNode][run]);
                }
            }
        }
    }
};

File: Python/Graphs/L_modifyGraphEdgeWeight.py
import heapq


def modifiedGraphEdges(self, n, edges, source, destination, target):
    adjacency_list = [[] for _ in range(n)]
    for i, (nodeA, nodeB, weight) in enumerate(edges):
        adjacency_list[nodeA].append((nodeB, i))
        adjacency_list[nodeB].append((nodeA, i))

    distances = [[float('inf')] * 2 for _ in range(n)]
    distances[source][0] = distances[source][1] = 0

    self.run_dijkstra(adjacency_list, edges, distances, source, 0, 0)
    difference = target - distances[destination][0]

    if difference < 0:
        return []

    self.run_dijkstra(adjacency_list, edges, distances, source, difference, 1)

    if distances[destination][1] < target:
        return []

    for edge in edges:
        if edge[2] == -1:
            edge[2] = 1

    return edges

def run_dijkstra(self, adjacency_list, edges, distances, source, difference, run):
    n = len(adjacency_list)
    priority_queue = [(0, source)]
    distances[source][run] = 0

    while priority_queue:
        current_distance, current_node = heapq.heappop(priority_queue)
        if current_distance > distances[current_node][run]:
            continue

        for next_node, edge_index in adjacency_list[current_node]:
            weight = edges[edge_index][2]
            if weight == -1:
                weight = 1
            if run == 1 and edges[edge_index][2] == -1:
                new_weight = difference + distances[next_node][0] - distances[current_node][1]
                if new_weight > weight:
                    edges[edge_index][2] = weight = new_weight

            if distances[next_node][run] > distances[current_node][run] + weight:
                distances[next_node][run] = distances[current_node][run] + weight
                heapq.heappush(priority_queue, (distances[next_node][run], next_node))
